compiler/p/codegen/OMRTreeEvaluator.cpp

```diff
@@ -4570,8 +4570,11 @@ TR::Register *OMR::Power::TreeEvaluator::arraycopyEvaluator(TR::Node *node, TR::
       {
       int64_t len = (lengthNode->getType().isInt32() ?
                      lengthNode->getInt() : lengthNode->getLongInt());
-      // for simple arraycopies the helper is better if length is long
-      if (len>=0 && (!simpleCopy || len < MAX_PPC_ARRAYCOPY_INLINE))
+
+      // inlineArrayCopy is not currently capable of handling very long lengths correctly. Under some circumstances, it
+      // will generate an li instruction with an out-of-bounds immediate, which triggers an assert in the binary
+      // encoder.
+      if (len>=0 && (!simpleCopy || len < MAX_PPC_ARRAYCOPY_INLINE) && len < 0x100000000LL)
```

Comment:
MAX_PPC_ARRAYCOPY_INLINE is intended as the inlining gating length. Beyond that length, inlined code likely performed worse than calling out to the helper, since the helper used wider load/store and tried to align at least the load side. So, the fix could simply be:  if (len >=0 && len < MAX_PPC_ARRAYCOPY_INLINE)
Make senses?

Need to revisit these conditions in the future with new processor capability ...

Reply:
I suspected the original code probably assumed !simpleCopy is not going to be long (but, of course, that assumption didn't hold in reality).

Reply:
I'm not so sure that the original code was making assumptions about the lengths of 5-child arraycopies, as the comment above this code specifically called out simple arraycopies as being slower than the helper at long lengths and didn't mention anything about any such assumption. Some corresponding code in OpenJ9 (see eclipse/openj9#8529) also doesn't seem to have any previous upper bound on length for inlining. That being said, you may very well be correct as our legacy code has been known to make assumptions like this without documenting it...

Regardless, I don't think trying to correct that at the moment would be a good idea. Making such a change carries a big risk of introducing performance regression in my opinion, since the current value of `MAX_PPC_ARRAYCOPY_INLINE` (256) seems like it would be commonly exceeded. Considering that naively backing this change out after the new binary encoder lands would cause spurious assertion failures in OpenJ9 due to eclipse/openj9#8530, I think that such a change should probably be made separately regardless. I suggest you open an issue to track this so that this can be investigated more in-depth at a later date.

Reply:
I am not worried about performance regression at all, since both simple and !simple copy(s) used the same code for the portion of memory movement. Rather, I am slightly concerned with opening up new functionality (inlined, that is), but overall not much. In addition, the typical java arraycopy length is under 64 bytes.

Reply:
After chatting more in-depth with @zl-wang regarding the reasoning for not using arraycopy inlining with long lengths, I'm now convinced that it would be best to restrict 5-child arraycopy nodes to the same maximum length as 3-child nodes. That has been implemented in d6ac306f1f3273f00775ad88eac660c93af285d5. I have, however, put this in a separate commit just in case there is some sort of performance issue that would require that to be reverted.